File: android/app/src/main/java/se/iloppis/app/data/RegisterSessionManager.kt
package se.iloppis.app.data

import android.content.Context
import androidx.core.content.edit
import se.iloppis.app.network.cashier.RegisterLifecycleEventType
import java.util.UUID

/**
 * Persists session state in SharedPreferences so it survives process death and restarts.
 * State machine mirrors the backend/desktop contract:
 *
 * ```
 * OPEN ──► CLOSE_REQUESTED ──► CLOSED
 * ```
 *
 * CLOSED is terminal; call [openSession] to start a fresh session.
 *
 * Thread-safety: public methods annotated with [Synchronized] are synchronized on this
 * [RegisterSessionManager] instance monitor.
 */
class RegisterSessionManager private constructor(private val appContext: Context) {

    enum class State { OPEN, CLOSE_REQUESTED, CLOSED }

    data class Session(
        val sessionId: String,
        val eventId: String,
        val registerId: String,
        val state: State,
        /**
         * The lifecycle event type that should be sent on the NEXT heartbeat tick after a
         * transition. When no transition is pending, [recordSync] can set SYNC to report
         * liveness once per successful heartbeat cycle.
         */
        val pendingLifecycleEvent: RegisterLifecycleEventType?
    )

    private val prefs by lazy {
        appContext.getSharedPreferences("register_session", Context.MODE_PRIVATE)
    }

    private var current: Session? = null

    init {
        current = loadFromPrefs()
    }

    /**
     * Replaces any current session with a new open session for the given event and register.
     * The returned session has a fresh ID and a pending `OPEN` lifecycle event.
     */
    @Synchronized
    fun openSession(eventId: String, registerId: String): Session {
        val s = Session(
            sessionId = UUID.randomUUID().toString(),
            eventId = eventId,
            registerId = registerId,
            state = State.OPEN,
            pendingLifecycleEvent = RegisterLifecycleEventType.REGISTER_LIFECYCLE_OPEN
        )
        current = s
        persist(s)
        return s
    }

    /**
     * Moves an open session to `CLOSE_REQUESTED` and schedules that lifecycle event.
     *
     * @return `true` only when the state transition was applied.
     */
    @Synchronized
    fun requestClose(): Boolean {
        val s = current ?: return false
        if (s.state != State.OPEN) return false
        val updated = s.copy(
            state = State.CLOSE_REQUESTED,
            pendingLifecycleEvent = RegisterLifecycleEventType.REGISTER_LIFECYCLE_CLOSE_REQUESTED
        )
        current = updated
        persist(updated)
        return true
    }

    /**
     * Schedules `CLOSE_CONFIRMED` for a session awaiting closure.
     * The session becomes closed only after [clearPendingLifecycleEvent] acknowledges it.
     *
     * @return `true` only when confirmation was scheduled.
     */
    @Synchronized
    fun confirmClose(): Boolean {
        val s = current ?: return false
        if (s.state != State.CLOSE_REQUESTED) return false
        val updated = s.copy(
            pendingLifecycleEvent = RegisterLifecycleEventType.REGISTER_LIFECYCLE_CLOSE_CONFIRMED
        )
        current = updated
        persist(updated)
        return true
    }

    /**
     * Clears a delivered lifecycle event only when both expected values still match.
     * This prevents a stale response from discarding a newer session transition.
     */
    @Synchronized
    fun clearPendingLifecycleEvent(
        expectedLifecycleEvent: RegisterLifecycleEventType?,
        expectedSessionId: String?
    ) {
        val s = current ?: return
        if (expectedLifecycleEvent == null || expectedSessionId.isNullOrBlank()) return
        if (s.sessionId != expectedSessionId) return
        if (s.pendingLifecycleEvent != expectedLifecycleEvent) return
        val nextState = if (
            expectedLifecycleEvent == RegisterLifecycleEventType.REGISTER_LIFECYCLE_CLOSE_CONFIRMED &&
            s.state == State.CLOSE_REQUESTED
        ) {
            State.CLOSED
        } else {
            s.state
        }
        current = s.copy(state = nextState, pendingLifecycleEvent = null)
        persist(current!!)
    }

    /** Schedules a liveness sync unless a transition is pending or the session is closed. */
    @Synchronized
    fun recordSync() {
        val s = current ?: return
        if (s.state == State.CLOSED) return
        if (s.pendingLifecycleEvent != null) return
        val updated = s.copy(
            pendingLifecycleEvent = RegisterLifecycleEventType.REGISTER_LIFECYCLE_SYNC
        )
        current = updated
        persist(updated)
    }

    /** Returns the current persisted session snapshot, or `null` before the first session. */
    @Synchronized
    fun getCurrent(): Session? = current

    /** Returns whether the current session can still send cashier activity. */
    @Synchronized
    fun isSessionActive(): Boolean =
        current?.state == State.OPEN || current?.state == State.CLOSE_REQUESTED

    private fun persist(s: Session) {
        val commitNow = s.state == State.CLOSED
        prefs.edit(commit = commitNow) {
            putString("session_id", s.sessionId)
            putString("event_id", s.eventId)
            putString("register_id", s.registerId)
            putString("state", s.state.name)
            putString("pending_lifecycle", s.pendingLifecycleEvent?.name)
        }
    }

    private fun loadFromPrefs(): Session? {
        val sessionId = prefs.getString("session_id", null) ?: return null
        val eventId = prefs.getString("event_id", null) ?: return null
        val registerId = prefs.getString("register_id", null) ?: return null
        val state = prefs.getString("state", null)?.let {
            runCatching { State.valueOf(it) }.getOrNull()
        } ?: return null
        val pending = prefs.getString("pending_lifecycle", null)?.let {
            runCatching { RegisterLifecycleEventType.valueOf(it) }.getOrNull()
        }
        return Session(sessionId, eventId, registerId, state, pending)
    }

    companion object {
        @Volatile
        private var instance: RegisterSessionManager? = null

        /** Returns the process-wide manager backed by the application context. */
        fun getInstance(context: Context): RegisterSessionManager =
            instance ?: synchronized(this) {
                instance ?: RegisterSessionManager(context.applicationContext).also { instance = it }
            }
    }
}

File: ios/iLoppis/UI/Screens/Cashier/RegisterSessionManager.swift
import Foundation

@MainActor
final class RegisterSessionManager {
    enum State: String {
        case open = "OPEN"
        case closeRequested = "CLOSE_REQUESTED"
        case closed = "CLOSED"
    }

    struct Session: Equatable {
        let sessionId: String
        let eventId: String
        let registerId: String
        let state: State
        let pendingLifecycleEvent: RegisterLifecycleEventType?
    }

    private let defaults: UserDefaults
    private let storageKey = "cashier_register_session"
    private var current: Session?

    init(defaults: UserDefaults = .standard) {
        self.defaults = defaults
        self.current = loadFromDefaults()
    }

    func openSession(eventId: String, registerId: String) -> Session {
        let session = Session(
            sessionId: UUID().uuidString,
            eventId: eventId,
            registerId: registerId,
            state: .open,
            pendingLifecycleEvent: .open
        )
        current = session
        persist(session)
        return session
    }

    func requestClose() -> Bool {
        guard let session = current, session.state == .open else {
            return false
        }
        let updated = Session(
            sessionId: session.sessionId,
            eventId: session.eventId,
            registerId: session.registerId,
            state: .closeRequested,
            pendingLifecycleEvent: .closeRequested
        )
        current = updated
        persist(updated)
        return true
    }

    func confirmClose() -> Bool {
        guard let session = current, session.state == .closeRequested else {
            return false
        }
        let updated = Session(
            sessionId: session.sessionId,
            eventId: session.eventId,
            registerId: session.registerId,
            state: session.state,
            pendingLifecycleEvent: .closeConfirmed
        )
        current = updated
        persist(updated)
        return true
    }

    func clearPendingLifecycleEvent(expectedLifecycleEvent: RegisterLifecycleEventType?, expectedSessionId: String?) {
        guard
            let expectedLifecycleEvent,
            let expectedSessionId,
            !expectedSessionId.isEmpty,
            let session = current,
            session.sessionId == expectedSessionId,
            session.pendingLifecycleEvent == expectedLifecycleEvent
        else {
            return
        }

        let nextState: State
        if expectedLifecycleEvent == .closeConfirmed && session.state == .closeRequested {
            nextState = .closed
        } else {
            nextState = session.state
        }

        let updated = Session(
            sessionId: session.sessionId,
            eventId: session.eventId,
            registerId: session.registerId,
            state: nextState,
            pendingLifecycleEvent: nil
        )
        current = updated
        persist(updated)
    }

    func recordSync() {
        guard let session = current, session.state != .closed, session.pendingLifecycleEvent == nil else {
            return
        }
        let updated = Session(
            sessionId: session.sessionId,
            eventId: session.eventId,
            registerId: session.registerId,
            state: session.state,
            pendingLifecycleEvent: .sync
        )
        current = updated
        persist(updated)
    }

    func getCurrent() -> Session? {
        current
    }

    func isSessionActive() -> Bool {
        current?.state == .open || current?.state == .closeRequested
    }

    private func persist(_ session: Session) {
        let payload: [String: String] = [
            "session_id": session.sessionId,
            "event_id": session.eventId,
            "register_id": session.registerId,
            "state": session.state.rawValue,
            "pending_lifecycle": session.pendingLifecycleEvent?.rawValue ?? ""
        ]
        defaults.set(payload, forKey: storageKey)
    }

    private func loadFromDefaults() -> Session? {
        guard
            let payload = defaults.dictionary(forKey: storageKey) as? [String: String],
            let sessionId = payload["session_id"], !sessionId.isEmpty,
            let eventId = payload["event_id"], !eventId.isEmpty,
            let registerId = payload["register_id"], !registerId.isEmpty,
            let stateRaw = payload["state"],
            let state = State(rawValue: stateRaw)
        else {
            return nil
        }

        let pendingRaw = payload["pending_lifecycle"] ?? ""
        let pendingLifecycle = pendingRaw.isEmpty ? nil : RegisterLifecycleEventType(rawValue: pendingRaw)

        return Session(
            sessionId: sessionId,
            eventId: eventId,
            registerId: registerId,
            state: state,
            pendingLifecycleEvent: pendingLifecycle
        )
    }
}
